documentation/specs/question.md

```diff
@@ -0,0 +1,84 @@
+
+MSBuild can skip Target or Task from running again by implementing some checks. Targets uses the Inputs and Outputs parameters to compare the timestamp of input and output files (see ['Build incrementally'](https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-build-incrementally) for details). Tasks have different behavior and thus have different rules. See below for details of each task shipped with MSBuild. Custom tasks can implement `IIncrementalTask` interface.
+
+Question switch ask if the next build is up-to-date. It will start a build, but will error out as soon as a target or task is not up-to-date. This error stops the build and allows investigation at the point of failure. It is recommended to use binlog logging to gather all the information. Targets or tasks that don't have an up-to-date check will build normally. Question is a check to help ensure minimal builds with zeros changes, and thus, faster build with small changes.
+
+[Fast Up-To-Date Check](https://github.com/dotnet/project-system/blob/cd275918ef9f181f6efab96715a91db7aabec832/docs/up-to-date-check.md) is a system that is implemented by the Project System, that decides, if it needs to run MSBuild.  MSBuild takes a non-trival amount of time to load, evaluate, and run through each target and task.  Fast Up-To-Date is faster, but can be less accurate, suitable for an IDE and a human interface. Question is not and does not replace Fast Up-To-Date Check.
+
+## Usage
+Question mode is designed to be used on the command line.  Run your normal build, then run again with /question.
+```
+msbuild /p:Configuration=Debug Project1.csproj /bl:build.binlog
+msbuild /p:Configuration=Debug Project1.csproj /bl:incremental.binlog /question
+```
+If there are no errors, then your build is up-to-date.  
+If there are errors, then investigate the error.  See common errors below.  Keep both logs to help with your investigation.
+
+## Custom Tasks
+Task author can implement the optional `IIncrementalTask` interface that will explose `FailIfNotIncremental`. `FailIfNotIncremental` is true when `/question` switch is used. The custom task will need to decide how it want to handle its behavior.  For example - if there is already a message describing why the task cannot be skipped, then simply convert the message to an error. Remember to return `false` to stop the build.  For the best reproducibility, try to preserve the input state and exit early.
+
+```C#
+if (FailIfNotIncremental)
+{
+  TaskLoggingHelper.LogErrorWithCodeFromResources("ToolTask.NotUpToDate");
+  return false;
+}
+else
+{
+  TaskLoggingHelper.LogMessageWithCodeFromResources("ToolTask.NotUpToDate");
+}
+```
+
+
+### `ToolTask`
+If inheriting from ToolTask, your custom task can override `SkipTaskExecution()`.  When it returns `false`, TookTask will exit with error.
+
+## Shipping Tasks
+When `/question` switch is used, it will modify the shipping task with these behavior.  Note: this is still experimental and can change.
```

Comment:
```suggestion
When `/question` switch is used, it will modify the shipping task behavior as follows.  Note: this is still experimental and can change.
```